### PR TITLE
Implemented RMSProp Algorithm

### PR DESCRIPTION
**Pull request by NivenT**
I added [RMSProp](http://sebastianruder.com/optimizing-gradient-descent/index.html#rmsprop) as another gradient descent algorithm. I think I have heard that it has become a popular optimization algorithm recently, so I thought it would nice to add. I tested it on the simple model in the tests folder, and on the neural network example.

### CODE REVIEW

**Comment by AtheMathmo** [comment]
This is really cool! Thanks for the thorough implementation.

I had a quick check through and it looks great. I'll take a more thorough look tomorrow and hopefully get this merged.

File: src/learning/optim/grad_desc.rs [hunk 1]
```diff
@@ -303,10 +303,121 @@ impl<M: Optimizable<Inputs = Matrix<f64>, Targets = Matrix<f64>>> OptimAlgorithm
...
+                                                      &inputs.select_rows(&[*i]),
+                                                      &targets.select_rows(&[*i]));
+
+                let grad = Vector::new(grad);
+                let grad_squared = grad.clone().apply(&|x| x*x);
+                // Update cached average of squared gradients
+                rmsprop_cache = &rmsprop_cache*self.decay_rate + &grad_squared*(1.0 - self.decay_rate);
+                // RMSProp update rule 
+                params = &params - (&grad*self.learning_rate).elediv(
+                                        &rmsprop_cache.clone().apply(&|x| (x + self.epsilon).sqrt()));
```
Comment (AtheMathmo) [inline src/learning/optim/grad_desc.rs]: You can avoid the clone here by using the `utils::vec_bin_op` from `rulinalg`. Something like:

`params = &params - utils::in_place_vec_bin_op((&grad * self.learning_rate).mut_data(), rmsprop_cache.data(), |x, &y| {*x =  *x * (y + self.epsilon).sqrt() })`

It's verbose but more efficient. Essentially you scale `grad` and then mutate this data without any cloning etc. It will also be vectorized etc.

**Comment by AtheMathmo** [comment]
I have one minor comment to help prevent some allocation - otherwise this looks good!

**Comment by NivenT** [comment]
Thanks for the suggestion. That clone was bothering me but I was not sure how to get rid of it.

**Comment by AtheMathmo** [comment]
I guess the rulinalg API needs a little refining so that it is more obvious how to achieve things like that...

**Comment by NivenT** [comment]
Maybe. To be completely honest, I did not spend a lot of time looking for a way to do it. How does it look now?

**Comment by AtheMathmo** [comment]
Looks good to me. I'm just going to try out the algorithm in the neural net example to double check that nothing odd happens. If it runs well I'll be happy to merge!

**Comment by AtheMathmo** [comment]
It looks good - thanks for this! RMSProp looks really cool :)
